src/pharmpy/expressions.py

```diff
@@ -56,3 +56,142 @@ def _free_images_iter(expr: sympy.Expr) -> Iterable[sympy.Expr]:
 def assume_all(predicate: sympy.assumptions.Predicate, expressions: Iterable[sympy.Expr]):
     tautology = sympy.Q.is_true(True)
     return reduce(__and__, map(predicate, expressions), tautology)
+
+
+def xreplace_dict(dictlike) -> Dict[sympy.Expr, sympy.Expr]:
+    return {_sympify_old(key): _sympify_new(value) for key, value in dictlike.items()}
+
+
+def _sympify_old(old) -> sympy.Expr:
+    # NOTE This mimics sympy's input coercion in subs
+    return (
+        sympy.Symbol(old)
+        if isinstance(old, str)
+        else sympy.sympify(old, strict=not isinstance(old, type))
+    )
+
+
+@lru_cache(maxsize=256)
+def _sympify_new(new) -> sympy.Expr:
+    # NOTE This mimics sympy's input coercion in subs
+    return sympy.sympify(new, strict=not isinstance(new, (str, type)))
+
+
+def subs(expr: sympy.Expr, mapping: Dict[sympy.Expr, sympy.Expr], simultaneous: bool = False):
```

Comment:
Do we need the `simultaneous` option? In my mind it should always be `True`, but perhaps there are use cases for `False` that I couldn't think of.

Reply:
I wanted this to be a drop-in replacement for `.subs`. Some current uses of `subs` seemed to need `simultaneous = False` but this can be revisited.

The advantage of having a `subs` function that can do everything correctly and some things faster than `.subs` is that it can be used everywhere without thinking. If we manage to have a faster `subs` even when `simultaneous` is `False` then we won't have to change anything to benefit from the speed-up.

Reply:
Another example where a fast non-simultaneous `subs` could be useful is when computing `full_expression`.